internal/daemon/controller/handlers/targets/target_service.go

```diff
@@ -234,87 +234,105 @@ func (s Service) ListTargets(ctx context.Context, req *pbs.ListTargetsRequest) (
+	// Use the requested page size only if it is smaller than
+	// the configured max.
+	if req.GetPageSize() != 0 && uint(req.GetPageSize()) < s.maxPageSize {
+		pageSize = int(req.GetPageSize())
+	}
+	filter, err := handlers.NewFilter(ctx, req.GetFilter())
 	if err != nil {
 		return nil, err
 	}
-	filter, err := handlers.NewFilter(ctx, req.GetFilter())
+
+	repo, err := s.repoFn(target.WithPermissions(userPerms))
 	if err != nil {
 		return nil, err
 	}
-	listTargetsFn := func(prevPageLast target.Target, refreshToken *pbs.ListRefreshToken, limit int) ([]target.Target, error) {
-		opts := []target.Option{
-			target.WithLimit(limit),
-		}
-		if prevPageLast == nil {
-			if refreshToken != nil {
-				opts = append(opts, target.WithStartPageAfterItem(
-					refreshToken.GetLastItemId(),
-					refreshToken.GetLastItemUpdatedTime().AsTime(),
-				))
-			}
-		} else {
-			opts = append(opts, target.WithStartPageAfterItem(prevPageLast.GetPublicId(), prevPageLast.GetUpdateTime().AsTime()))
+	filterItemFn := func(item target.Target) (bool, error) {
+		pbItem, err := toProto(ctx, item, newOutputOpts(ctx, item, authResults, authzScopes)...)
+		if err != nil {
+			return false, err
 		}
-		return repo.ListTargets(ctx, opts...)
-	}
-	filterAndConvertFn := func(item target.Target) (*pb.Target, error) {
-		pr := perms.Resource{Id: item.GetPublicId(), ScopeId: item.GetProjectId(), Type: resource.Target}
-		outputFields := authResults.FetchOutputFields(pr, action.List).SelfOrDefaults(authResults.UserId)
 
-		outputOpts := make([]handlers.Option, 0, 3)
-		outputOpts = append(outputOpts, handlers.WithOutputFields(outputFields))
-
-		if outputFields.Has(globals.ScopeField) {
-			outputOpts = append(outputOpts, handlers.WithScope(authzScopes[item.GetProjectId()]))
-		}
-		if outputFields.Has(globals.AuthorizedActionsField) {
-			authorizedActions := authResults.FetchActionSetForId(ctx, item.GetPublicId(), IdActions, auth.WithResource(&pr)).Strings()
-			outputOpts = append(outputOpts, handlers.WithAuthorizedActions(authorizedActions))
+		filterable, err := subtypes.Filterable(pbItem)
+		if err != nil {
+			return false, err
 		}
+		return filter.Match(filterable), nil
+	}
+	grantsHash, err := authResults.GrantsHash(ctx)
+	if err != nil {
+		return nil, err
+	}
 
-		pbItem, err := toProto(ctx, item, outputOpts...)
+	var listResp *target.ListResponse
+	if req.GetRefreshToken() == "" {
+		listResp, err = target.List(ctx, repo, grantsHash, pageSize, filterItemFn)
 		if err != nil {
 			return nil, err
 		}
-
-		filterable, err := subtypes.Filterable(pbItem)
+	} else {
+		rt, err := handlers.ParseRefreshToken(ctx, req.GetRefreshToken())
 		if err != nil {
 			return nil, err
 		}
-		if filter.Match(filterable) {
-			return pbItem, nil
+		// We're doing the conversion from the protobuf types to the
+		// domain types here rather than in the domain so that the domain
+		// doesn't need to know about the protobuf types.
+		domainRefreshToken := &refreshtoken.Token{
```

Comment:
You have a New(...) factory, but it's not used here?

Reply:
They are used for different purposes, the New factory is only used when we have a resource. I personally dislike constructors but I think we could make a constructor for this purpose and call it `New`, and in the existing use case for `New` we'll instead rename that function to `FromResource`.